### PR TITLE
RavenDB-18199 reorganize use of shard executor for BatchHandler & DocumentHandler

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18199

### CODE REVIEW

File: src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs [hunk 1]
```diff
@@ -11,14 +14,33 @@ namespace Raven.Server.Documents.Sharding.Executors
...
             _databaseContext = databaseContext;
+            var record = _databaseContext.DatabaseRecord;
+            _fullRange = Enumerable.Range(0, record.Shards.Length).ToArray();
+
+            _requestExecutors = new RequestExecutor[record.Shards.Length];
+            for (int i = 0; i < record.Shards.Length; i++)
+            {
+                var allNodes = store.GetClusterTopology().AllNodes;
+                var urls = record.Shards[i].AllNodes.Select(tag => allNodes[tag]).ToArray();
+                _requestExecutors[i] = RequestExecutor.Create(
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs]: CreateForServer

File: src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs [hunk 2]
```diff
@@ -11,14 +14,33 @@ namespace Raven.Server.Documents.Sharding.Executors
...
+                    urls,
+                    ShardHelper.ToShardName(_databaseContext.DatabaseName, i),
+                    store.Server.Certificate.Certificate,
+                    new DocumentConventions());
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs]: DocumentConventions.DefaultForServer

File: src/Raven.Server/Documents/Sharding/Operations/FetchDocumentsFromShardsOperation.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                {
+                    BlittableJsonReaderObject.PropertyDetails prop = default;
+                    for (int i = 0; i < cmdIncludes.Count; i++)
+                    {
+                        cmdIncludes.GetPropertyByIndex(i, ref prop);
+                        includesMap[prop.Name] = ((BlittableJsonReaderObject)prop.Value).Clone(_context);
+                    }
+                }
+
+                var positions = _idsByShards.ElementAt(shardNumber).Value.Positions;
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/Operations/FetchDocumentsFromShardsOperation.cs]: Huh? You are accessing a dictionary by numeric index?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 4]
```diff
@@ -859,8 +859,12 @@ protected virtual async Task<bool> TryLoadFromCacheAsync(JsonOperationContext co
                     return;
 
                 var refreshTask = RefreshIfNeeded(chosenNode, response);
+                
                 command.StatusCode = response.StatusCode;
 
+                if (response.IsSuccessStatusCode || command.StatusCode == HttpStatusCode.NotModified)
+                    command.Etag = response.GetEtagHeader() ?? cachedChangeVector;
+
```
Comment (ayende) [inline src/Raven.Client/Http/RequestExecutor.cs:867]: Why is this needed?

File: src/Raven.Client/Http/RavenCommand.cs [hunk 5]
```diff
@@ -66,6 +66,12 @@ public abstract class RavenCommand<TResult>
 
         internal long FailoverTopologyEtag = -2;
 
+        internal string Etag; // for NotModified use
+
+        internal Action<HttpRequestMessage> ModifyRequest;
+
+        internal Func<string, string> ModifyUrl;
+
```
Comment (ayende) [inline src/Raven.Client/Http/RavenCommand.cs:74]: Why is this needd?

File: src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs]: done

File: src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Executors/ShardExecutor.cs]: done

File: src/Raven.Server/Documents/Sharding/Operations/FetchDocumentsFromShardsOperation.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Operations/FetchDocumentsFromShardsOperation.cs]: removed

File: src/Raven.Client/Http/RavenCommand.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Client/Http/RavenCommand.cs:74]: we need a way to pass the some headers to all shards.. 
e.g. LastKnownRaftLast header

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Client/Http/RequestExecutor.cs:867]: to handle the `NotModified` on the orchestrator level
